### ISSUE TITLE
Reproducing FB15K Results / FB15K-237?

### ISSUE THREAD

**Issue by drcdr** [baharefatemi/SimplE#2]
I am trying to reproduce results, as well as try a new dataset, FB15K-237; I'd appreciate any thoughts you have (see questions at the end):

**Reproducing FB15K results**: I am using Windows10, pytorch-nightly (1.1, May15), TitanXP.  I ran the example from the README.md:

```python main.py -ne 1000 -lr 0.05 -reg 0.1 -dataset FB15K -emb_dim 200 -neg_ratio 10 -batch_size 4832 -save_each 50```

Training took right about 6 hours.  Validation takes 1250 seconds/epoch.  Results:

Source | FB15K | MRR Filter | MRR Raw | Hit@1 | Hit@3 | Hit@10
-- | -- | -- | -- | -- | -- | --
Paper | SimplE-ignr | 0.700 | 0.237 | 0.625 | 0.754 | 0.821
Paper | SimplE | 0.727 | 0.239 | 0.660 | 0.773 | 0.838
My Results | SimplE-a | 0.726 | 0.240 | 0.659 | 0.770 | 0.837

**FB15K-237**
I looked at [RotatE at github](https://github.com/DeepGraphLearning/KnowledgeGraphEmbedding/tree/master/data), and noticed that the test.txt, train.txt, and valid.txt exactly matched the files in your FB15K directory. So, I thought I would be able to run their FB15k-237 dataset with SimplE.  Using the same command as above (but referencing FB15K237), I got the following, which seems a little low (notwithstanding this is a much harder dataset, with inverse relations removed):

Source | FB15K237 | MRR Filter | MRR Raw | Hit@1 | Hit@3 | Hit@10
-- | -- | -- | -- | -- | -- | --
My Results | SimplE-b | 0.168 | 0.074 | 0.094 | 0.178 | 0.324

**Questions:**
1) What is the proper command to recreate the results for SimplE, from the paper?  Already, I can see I should have used 0.03 for the learning rate, for example. But clearly, the results are quite close anyway.
2) Are the result from the paper the result of a single run, or the average/best over N runs?
3) Are the modifications needed to run "SimplE-ignr" straightforward, or more involved?
4) Are my run-times about the same as what you get?
5) Does one need to do anything special to incorporate a new dataset like FB15K-237?  The train/test/val files look like the same format to me.
6) Does the removal of inverse relations from the dataset impact SimpleE?  Based on some of your comments in the paper, it seems it might, but where in the code is the impact particularly felt?
7) Regarding section 6.2 from the paper, how did you incorporate the background knowledge?

Thanks!

**Comment by susheels** [comment]
Hi,
I also did some experiments with FB15K-237 (inverse relations removed) for SimplE. 
I used the OpenKE-PyTorch implementation. Here are the results I got. For comparison I have RotatE results as well from their paper. 

No Type constraints are used for evaluation.
Model Parameters:

- Learning Rate - 0.02
- Embedding Dim - 200
- Batch Size - 2048

Model | MR(Filter)|MRR (Filter) | MRR (Raw) | Hit@10 | Hit@3 | Hit@1
-- | -- | -- | -- | -- | -- |--
OpenKE (SimplE-ignr)|299|0.261|0.143|0.471|0.304|0.155
RotatE (paper results)|177| 0.338| -| 0.533|0.375|0.241

So looks like SimplE does suffer when inverse relations are removed, but the results I get are better than @drcdr got, maybe it's the learning rate. Also, only while testing I don't use inverse relations embedding (ignr method is used).

Hope this info is useful!

Thanks!

**Comment by baharefatemi** [comment]
> I am trying to reproduce results, as well as try a new dataset, FB15K-237; I'd appreciate any thoughts you have (see questions at the end):
> 
> **Reproducing FB15K results**: I am using Windows10, pytorch-nightly (1.1, May15), TitanXP. I ran the example from the README.md:
> 
> `python main.py -ne 1000 -lr 0.05 -reg 0.1 -dataset FB15K -emb_dim 200 -neg_ratio 10 -batch_size 4832 -save_each 50`
> 
> Training took right about 6 hours. Validation takes 1250 seconds/epoch. Results:
> 
> Source	FB15K	MRR Filter	MRR Raw	Hit@1	Hit@3	Hit@10
> Paper	SimplE-ignr	0.700	0.237	0.625	0.754	0.821
> Paper	SimplE	0.727	0.239	0.660	0.773	0.838
> My Results	SimplE-a	0.726	0.240	0.659	0.770	0.837
> **FB15K-237**
> I looked at [RotatE at github](https://github.com/DeepGraphLearning/KnowledgeGraphEmbedding/tree/master/data), and noticed that the test.txt, train.txt, and valid.txt exactly matched the files in your FB15K directory. So, I thought I would be able to run their FB15k-237 dataset with SimplE. Using the same command as above (but referencing FB15K237), I got the following, which seems a little low (notwithstanding this is a much harder dataset, with inverse relations removed):
> 
> Source	FB15K237	MRR Filter	MRR Raw	Hit@1	Hit@3	Hit@10
> My Results	SimplE-b	0.168	0.074	0.094	0.178	0.324
> **Questions:**
> 
> 1. What is the proper command to recreate the results for SimplE, from the paper?  Already, I can see I should have used 0.03 for the learning rate, for example. But clearly, the results are quite close anyway.
> 2. Are the result from the paper the result of a single run, or the average/best over N runs?
> 3. Are the modifications needed to run "SimplE-ignr" straightforward, or more involved?
> 4. Are my run-times about the same as what you get?
> 5. Does one need to do anything special to incorporate a new dataset like FB15K-237?  The train/test/val files look like the same format to me.
> 6. Does the removal of inverse relations from the dataset impact SimpleE?  Based on some of your comments in the paper, it seems it might, but where in the code is the impact particularly felt?
> 7. Regarding section 6.2 from the paper, how did you incorporate the background knowledge?
> 
> Thanks!

**Comment by baharefatemi** [comment]
Hi @drcdr, I'm so sorry for my late response.
Just note that I'm not an author in this paper. The original implementation is in Tensorflow so I decided to implement this model in PyTorch. I was able to reproduce the results in the paper. 

1. I have the proper hyperparameters to get the results in the paper listed in the Github page.  I copy it here for you: python main.py -ne 1000 -lr 0.05 -reg 0.1 -dataset FB15K -emb_dim 200 -neg_ratio 10 -batch_size 4832 -save_each 50
2. The results I have here (which is quite similar to the ones in the paper) are for one single run. However, I noticed that if your run it multiple times, you get quite similar results.
3. The modifications to run SimplE-ignr should be straightforward. You basically need to ignore the inverse of relations at test time.
4. Your run time sounds reasonable. I haven't record mine to compare it with my yours seems reasonable.
5. For a new dataset, you should tune the hyperparameters yourself. The hyperparameters I wrote in the main page are for WN18 and Fb15K. If you want to work on a new dataset, you should tune all hyperparameters from scratch.
6 and 7. As I mentioned above, I'm not an author and I'd better not comment on this.
